warehouse/admin/views/blacklist.py

```diff
@@ -0,0 +1,251 @@
+                      .filter(Release.name == project.name)
+                      .all()
+        )
+        files = (
+            request.db.query(File)
+                      .filter(File.name == project.name)
+                      .all()
+        )
+        users = [
+            r.user
+            for r in (
+                request.db.query(Role)
+                .join(User)
+                .filter(Role.project == project)
+                .distinct(User.username)
+                .order_by(User.username)
+                .all()
+            )
+        ]
+    else:
+        releases = []
+        files = []
+        users = []
+
+    return {
+        "blacklist": {
+            "project": project_name,
+            "comment": comment,
+        },
+        "existing": {
+            "project": project,
+            "releases": releases,
+            "files": files,
+            "users": users,
+        }
+    }
+
+
+@view_config(
+    route_name="admin.blacklist.add",
+    permission="admin",
+    request_method="POST",
+    uses_session=True,
+    require_methods=False,
+)
+def add_blacklist(request):
+    project_name = request.POST.get("project")
+    if project_name is None:
+        raise HTTPBadRequest("Must have a project to confirm.")
+    comment = request.POST.get("comment", "")
+
+    # Verify that the user has confirmed the request to blacklist.
+    confirm = request.POST.get("confirm")
+    if not confirm:
+        request.session.flash(
+            "Must confirm the blacklist request.",
+            queue="error",
+        )
+        return HTTPMovedPermanently(request.current_route_path())
+    elif canonicalize_name(confirm) != canonicalize_name(project_name):
+        request.session.flash(
+            f"{confirm!r} is not the same as {project_name!r}",
+            queue="error",
+        )
+        return HTTPMovedPermanently(request.current_route_path())
+
+    # Add our requested blacklist.
+    request.db.add(
+        BlacklistedProject(
+            name=project_name,
+            comment=comment,
+            blacklisted_by=request.user,
+        )
+    )
+
+    # Go through and delete anything that we need to delete so that our
+    # blacklist actually blocks things and isn't ignored (since the blacklist
+    # only takes effect on new project registration).
+    # TODO: This should be in a generic function somewhere instead of putting
+    #       it here, however this will have to do for now.
+    # TODO: We don't actually delete files from the data store. We should add
+    #       some kind of garbage collection at some point.
+    project = (
+        request.db.query(Project)
+                  .filter(Project.normalized_name ==
+                          func.normalize_pep426_name(project_name))
+                  .first()
+    )
+    if project is not None:
+        request.db.add(
+            JournalEntry(
+                name=project.name,
+                action="remove",
```

Comment:
should the action here be `blacklisted`?

Reply:
I'm going to leave this alone, because other things might be reading our journal entries and inferring things from them, and we want them to still go ahead and remove these projects if they are. It's not really super important to keep in the log that a project was blacklisted either, since it's a binary state and admins can see when it was and who did it (as well as any relevant comments).